components/pedigree/resources/src/main/resources/pedigree/externalIdManager.js

```diff
@@ -0,0 +1,154 @@
+                }
+            }
+            if (externalID && externalID.length > 0) {
+                // update all the data structures
+                this._add(externalID, phenotipsID);
+                this.patientRecordToID[phenotipsID] = externalID;
+            }
+        },
+
+        remove: function(phenotipsID, removedFromPedigree)
+        {
+            if (this.patientRecordToID.hasOwnProperty(phenotipsID)) {
+                var oldID = this.patientRecordToID[phenotipsID];
+                Helpers.removeFirstOccurrenceByValue(this.externalIDs[oldID], phenotipsID);
+                if (this.externalIDs[oldID].length == 0) {
+                    delete this.externalIDs[oldID];
+                }
+                delete this.patientRecordToID[phenotipsID];
+
+                // when a patient is removed from the pedigree, its external ID may now be present outside of pedigree
+                if (removedFromPedigree) {
+                    Helpers.removeFirstOccurrenceByValue(this.externallyAbsentIDs, oldID);
+                }
+            }
+        },
+
+        isUniqueID: function(phenotipsID, candidateID, callbackOnYes, callbackOnNo)
+        {
+            if (candidateID.length == 0) {
+                return true;
+            }
+
+            if (this.externalIDs.hasOwnProperty(candidateID)) {
+                var recordsWithThisID = this.externalIDs[candidateID];
+                if (recordsWithThisID.length > 1) {
+                    callbackOnNo();
+                    return;
+                }
+                if (recordsWithThisID.length == 1 && recordsWithThisID[0] != phenotipsID) {
+                    callbackOnNo();
+                    return;
+                }
+                callbackOnYes();
+                return;
+            }
+
+            if (this.externallyAbsentIDs.hasOwnProperty(candidateID)) {
+                callbackOnYes();
+                return;
+            }
+
+            var onInvalid = function() {
+                // add it to the list of known taken IDs to avoid sending the same request again
+                this._add(candidateID, this._EXTERNAL_PATIENT);
+                callbackOnNo();
+            }.bind(this);
+
+            var onValid = function() {
+                this.externallyAbsentIDs[candidateID] = true;
+                callbackOnYes();
+            }.bind(this);
+
+            this._checkExternalIDinPhenotips(phenotipsID, candidateID, onValid, onInvalid);
+        },
+
+        duplicateIDPresent: function()
+        {
+            for(var id in this.externalIDs) {
+                if (this.externalIDs.hasOwnProperty(id)) {
+                    if (this.externalIDs[id].length > 1) {
+                        var involvesExternalPatient = false;
+                        if (Helpers.arrayContains(this.externalIDs[id], this._EXTERNAL_PATIENT)) {
+                            involvesExternalPatient = true;
```

Comment:
`var involvesExternalPatient = Helpers.arrayContains(this.externalIDs[id], this._EXTERNAL_PATIENT);`